flutter-idea/src/io/flutter/run/daemon/DevToolsServerTask.java

```diff
@@ -0,0 +1,399 @@
+import io.flutter.dart.DtdUtils;
+import io.flutter.sdk.FlutterSdk;
+import io.flutter.sdk.FlutterSdkUtil;
+import io.flutter.utils.JsonUtils;
+import io.flutter.utils.MostlySilentColoredProcessHandler;
+import io.flutter.utils.OpenApiUtils;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+class DevToolsServerTask extends Task.Backgroundable {
+  private static final Logger LOG = Logger.getInstance(DevToolsServerTask.class);
+  public static final String LOCAL_DEVTOOLS_DIR = "flutter.local.devtools.dir";
+  public static final String LOCAL_DEVTOOLS_ARGS = "flutter.local.devtools.args";
+  private @NotNull Project project;
+  private final AtomicReference<CompletableFuture<DevToolsInstance>> devToolsFutureRef;
+
+  private DaemonApi daemonApi;
+  private ProcessHandler process;
+
+  public DevToolsServerTask(@NotNull Project project,
+                            @NotNull String title,
+                            AtomicReference<CompletableFuture<DevToolsInstance>> devToolsFutureRef) {
+    super(project, title, true);
+    this.project = project;
+    this.devToolsFutureRef = devToolsFutureRef;
+  }
+
+  @Override
+  public void run(@NotNull ProgressIndicator progressIndicator) {
+    try {
+      progressIndicator.setFraction(30);
+      progressIndicator.setText2("Init");
+
+      // If DevTools is not supported, start the daemon instead.
+      final boolean dartDevToolsSupported = dartSdkSupportsDartDevTools();
+      if (!dartDevToolsSupported) {
+        progressIndicator.setFraction(60);
+        progressIndicator.setText2("Daemon set-up");
+        setUpWithDaemon();
+        return;
+      }
+
+      // If we are in a Bazel workspace, start the server.
+      // Note: This is only for internal usages.
+      final WorkspaceCache workspaceCache = WorkspaceCache.getInstance(project);
+      if (workspaceCache.isBazel()) {
+        progressIndicator.setFraction(60);
+        progressIndicator.setText2("Starting server");
+        setUpWithDart(createCommand(workspaceCache.get().getRoot().getPath(), workspaceCache.get().getDevToolsScript(),
+                                    ImmutableList.of("--machine")));
+        return;
+      }
+
+      // This is only for development to check integration with a locally run DevTools server.
+      // To enable, follow the instructions in:
+      // https://github.com/flutter/flutter-intellij/blob/master/CONTRIBUTING.md#developing-with-local-devtools
+      final String localDevToolsDir = Registry.stringValue(LOCAL_DEVTOOLS_DIR);
+      if (!localDevToolsDir.isEmpty()) {
```

Comment:
Maybe we can add this to the log too? (May help with debugging down the road.) I'm thinking `LOG.info(...)`.

Reply:
Done